fletcher/base.py

```diff
@@ -703,28 +709,53 @@ def take(self, indices, allow_fill=False, fill_value=None):
             When `indices` contains negative values other than ``-1``
             and `allow_fill` is True.
 
-        Notes
-        -----
-        ExtensionArray.take is called by ``Series.__getitem__``, ``.loc``,
-        ``iloc``, when `indices` is a sequence of values. Additionally,
-        it's called by :meth:`Series.reindex`, or any other method
-        that causes realignemnt, with a `fill_value`.
-
         See Also
         --------
         numpy.take
         pandas.api.extensions.take
         """
-        from pandas.core.algorithms import take
+        if isinstance(indices, list):
+            indices = np.asarray(indices)
+            if not np.any(indices < 0):
+                allow_fill = False
+        if isinstance(indices,slice) and allow_fill:
+            Raise Exception('allow_fill cannot be used with slice')
+
+        if not allow_fill:
+            indices = self._indices_to_numpy_array(indices)
+            lengths = np.fromiter(map(len, self.data.chunks), dtype=np.int64)
+            cum_lengths = lengths.cumsum()
+
+            bins = np.searchsorted(cum_lengths, indices, side='right')  #one might want to parallel this operation
+            cum_lengths -= lengths
+            limits_idx = np.concatenate([[0], np.bincount(bins, minlength=self.data.num_chunks).cumsum()])
+
+            if is_increasing(bins):
+                sort_idx = None
```

Comment:
we don't need to sort anything if there's only one chunk, we can go directly for `pa.array.take`.

Reply:
I have put a special case when there is only one chunk where `pa.array.take` is used.